pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -451,7 +451,10 @@ func (c *MaxPDVolumeCountChecker) predicate(pod *v1.Pod, meta algorithm.Predicat
 		// violates MaxEBSVolumeCount or MaxGCEPDVolumeCount
 		return false, []algorithm.PredicateFailureReason{ErrMaxVolumeCountExceeded}, nil
 	}
-
+	// Computing them here for lack of better way to share things between predicates and priorities.
+	// TODO: @ravig Remove this.
+	nodeInfo.AllocatableVolumes = c.maxVolumes - numExistingVolumes
```

Comment:
This volume predicate only counts for 3 volume types, and only one of them can run in a single cluster.
- GCE PD
- AWS EBS
- Azure disk

Did you also want to count mounts for other volume types not included here, like NFS, Gluster, Ceph?

Also, the limits here are more about how many disks can be attached to a single node, not how many could be mounted.  So if you had multiple pods sharing the same volume on the same node, this predicate would only count 1 of them, even though there will be a mount for each pod.  But multiple pods sharing the same volume on a single node is not the common use case.

Reply:
In the future, it may be possible that you can have multiple volume types in the same cluster that each have their own max limit as well.

Reply:
Thanks for the information @msau42. I thought we are going in the direction of using a configmap which consists of max number of volumes per node based on cloud provider and CSI plugin, after which we can use configmap to get the information.

Reply:
I think the main challenge is that this predicate is only meant to restrict max number of attached volumes, not overall number of mounts.  If you want to spread number of attached volumes, then this predicate will work.

But it sounds like you want to spread out number of mounts, including for non-attachable plugins.

If that's the case, then you may need to consider a different method of counting mounts, such as just counting the total number of volumes in assigned pods.  But that still won't be completely accurate because each volume type may create a different number of mounts, like I mentioned earlier.

Reply:
> But it sounds like you want to spread out number of mounts, including for non-attachable plugins.

So, we wanted to find nodes that have balanced number of volumes here. Not sure where this PR gave the impression of spreading mounts. While finding node that has balanced mounts will be even more useful, I think it has to be dependent on lower level API talking to cloud provider or listing /etc/mtab in case of linux etc and is beyond the scope of what we want to do here.

Reply:
Ok, so to be clear, if you want to spread number of attached volumes to each node, then using this predicate will work, and we are working on enhancing this predicate in the future to also accommodate nodes with different limits, and also be more modular for other volume types.  cc @gnufied

Reply:
Thanks, @msau42! The goal here is to balance the number of attached volumes, not the mount points, but that could be another idea to explore.
As far as I know, the limit on the number of mount points is very large (I don't know the number, but something in the order thousands). If this is true, the number of mount points is practically unlimited for us. So, should we care about balancing the number of mount points?

Reply:
Yeah the limit of mount points is much larger than the limit of attachable volumes, so I agree balancing on attachable limit is more important.  I think I got confused because the original issue mentioned mount points.  Maybe improving the wording in the comments here to explicitly say we're balancing attachable volume cpunts would make it more clear.

Reply:
> Maybe improving the wording in the comments here to explicitly say we're balancing attachable volume cpunts would make it more clear.

Apologies if the comments are causing confusion. I thought I made it clear atleast from the start of PR, if you think otherwise, I can edit the existing comments. Please let me know, where the confusion is. Even if it is in code, I am happy to modify it.

Reply:
Maybe we can be more explicit in some of the variable names.  Like, instead of AllocableVolumes, something like AttachableVolumesLimit.

Reply:
I don't think, AttachableVolumesLimit actually reflects the allocatable volumes on the node. My intention there was to have a variable which holds the value for number of allocatable volumes currently on the node. AttachableVolumesLimit may be more suited for c.maxVolumes. I don't have strong opinion on it though.

Reply:
Oh sorry, yes, what about something like "CurrentAttachedVolumesCount"?  I want to be explicit that we are talking about attachable counts here.  There is another completely unrelated volume feature that uses the term "allocatable", so it would be better to align with the terminology that we use in the documentation and API fields.

Reply:
Sure. I can do that.